llvm/lib/AsmParser/LLLexer.cpp

```diff
@@ -737,6 +737,7 @@ lltok::Kind LLLexer::LexIdentifier() {
   KEYWORD(live);
   KEYWORD(dsoLocal);
   KEYWORD(canAutoHide);
+  KEYWORD(importAsDec);
```

Comment:
Not too strong preference on the spelling but could be something like "importAs: dec" vs "importAs: def" (in case there is another variation later like "summary"). Otherwise we can consider that later.

Reply:
IIUC your point is the new field in `struct GVFlags` should carry 3 states in the field, `Definition`, `Declaration` and `Summary` since there is a known use case for import summaries.

I can see two different ways to do that
1. Rename `importAsDec` to `importType` and make it two bits.
2. Keep the current way to represent def vs decl. For the import summary use case, add another one-bit field.

I like 1 better since join two 1-bit fields to represent 3 states is not very natural. If this sounds good, I could make the change.

Reply:
Yes, my feeling is 1 would be more natural too. Thanks!

Reply:
Remind me - what is the summary import case?

Reply:
The use case is to propagate more function and param attributes across modules (beyond the ones currently in `FFlags`, and usually requiring knowledge of the body vs the call edges, unlike `NoRecurse`). The main one I'm interested in is the `initializes` one under review (https://github.com/llvm/llvm-project/pull/84803) but in general a few other attributes inferred by FunctionAttrs pass like readonly, nocapture, ... Not more complex and uninteresting attributes like `byval<ty>`.

We found that at least pre-link, there would need to be some info recorded about whether there are "interesting" attributes to propagate to trigger this fallback for too large or noinline cases. Otherwise, always importing the  fallback would be too much (from Mingming's measurements). Options for info to record pre-link considered:

1. subset of interesting attributes themselves (e.g., a subset of the hash-consed AttributeList)
2. or, a flag that just indicates there are "enough" interesting attributes as a hint for the indexing step

Some pros for option 1 and import summary
- we could import the summary instead as an even lighter weight option than import dec -- the summary holds the info so we don't need the IR. One FDO-based experiment on a large app: about 15% more function summary copies and 30% more "from" modules summaries. This is when we consider `initializes`, FFlags `ReadOnly`, `ReadNone`,  and `NoRecurse` as interesting attributes. I skipped `NoUnwind` since the app is compiled with `-fno-exceptions` and a lot of functions would have `NoUnwind`. May increase if we add other param attributes.
- may eventually help handle vague linkage attribute inference to handle a de-refinement problem in a lighter weight manner (intersecting summary-only attributes and propagating) than an alternative of importing the prevailing definition's bodies. We have some discussion of this elsewhere.

Some pros for option 2 and just import decl
- simpler summary
- simplicity of one fallback option instead of two

Reply:
>  Otherwise, always importing the fallback would be too much (from Mingming's measurements)

To add some details, I printed the number of `ImportList` map entries after [computeImportForModule](https://github.com/llvm/llvm-project/blob/93e2a9ab1b28e1b89d30b89dc2d5bb6f8cc66dc9/llvm/lib/Transforms/IPO/FunctionImport.cpp#L1005C10-L1005C32) in two settings
1) The default ThinLTO + FDO build
2) The same option as 1, plus adding {noinline, too-large} callee (marking it as declaration type) into the `ImportList` map.

The ratio of map entry is 1 : 2.5, meaning the fallback approach will increase the memory usage of indexing step and eat up a lot of resources. 

Thereby selective import approach is chosen
1. For more complete call-graph, module-summary-analysis pass can retrieve the set of callees from `!prof` metadata
2. To know the function attributes of function declarations (which may not exist on a function declaration but exist on function definitions), a subset of interesting attributes will be serialized in the per-module IR. Postlink pipeline can read the summary and set the function attributes.

Reply:
I revised the state to a 2-bit struct member of `GlobalValueSummary::GVFlags`

Reply:
Without either a definition or declaration in the importing module, how do we even know what symbol a summary and its attributes apply to and what do we do with them? Trying to understand the mechanics of how a "summary" import works before we reserve a flag bit for that case. Couldn't we just mark it as a declaration import which would essentially be a no-op if a declaration/def already exists in the module?

Reply:
For the summary import case, I'm mostly looking at direct call cases, so the importing module does have a declaration already, but the declaration doesn't have the attributes that the source module has (inferred from the definition and possibly transitively from callees definitions), so the summary helps update the decl [e.g., in thinLTOFinalizeInModule](https://github.com/llvm/llvm-project/blob/3009228a09dbfe04e0911fc19813ec72d389bc45/llvm/lib/Transforms/IPO/FunctionImport.cpp#L1415)

One could also mark it as a declaration import, but we'd want it to overwrite instead of no-op.

Other pros / cons for import summary vs import decl in earlier comment but happy to elaborate more or if more data is needed. Thanks!

Reply:
> Trying to understand the mechanics of how a "summary" import works before we reserve a flag bit for that case. Couldn't we just mark it as a declaration import which would essentially be a no-op if a declaration/def already exists in the module?

Previously I thought reserving one bit is necessary for postlink to tell one of three scenarios from {definition, declaration, summary}. However now after f2f discussion I think one-bit (just marking one of {definition, declaration}) should work.

To summarize our discussion, one-bit approach works for summary import in a selective manner.

1. The GVFlags could reserve one-bit for ImportType, and making it just declaration or definition.
2. In the indexing step, 'ComputeImportForModule' creates an entry for declaration (when definition import is too costly, etc) if call-graph needs it, or when interesting attributes exists.
3. The postlink optimizer sees IR and combined summary. If definition or declaration is already in the module (which is the case except for indirect callee, since IR verifier requests direct-callee should at least have a declaration), `FunctionImporter` could just read interesting attributes from the summary and set it on the function (def or decl). For the set of summaries without a function (the case for indirect callee for better call-graph sort),  IRMover will be updated in a subsequent patch to import the declaration. 
   * I prototyped the IRMover change. The regression test [shows](https://github.com/llvm/llvm-project/compare/main...minglotus-6:llvm-project:irlinker#diff-5c849764b0a8c2eb1062dc219405316161e757745b67f7aa31f101d0cdca5ed0R32-R34) large indirect callee declaration is imported along with attributes. 

I'm going to update this PR to use one-bit approach.

Reply:
Basically, one-bit serves the purpose to include the global-value-summary entry in the postlink combined summary, and indexing step can selectively set the bit since it has visibility of the combined summary. 

If the one-bit indicates definition, postlink will import function definition along with attributes if function doesn't exist. If the one-bit indicates declaration, postlink can annotate attributes according to summaries if declaration already exists, and use IRMover to import declarations that don't exist yet.

Reply:
Ah I see, yes that makes sense.

Reply:
Update to use one-bit state and the enum class [comment](https://github.com/llvm/llvm-project/pull/87597/files#diff-55dc08bb988d556731f68efe4a6d7c40b6cf4acc5a7dd4943ce335f89ce3fea7R435-R446). PTAL, thanks!

Re test coverage
- ThinLTO/X86/dot-dumpler.ll covers ModuleSummaryIndex::exportToDot
- Transforms/WholeProgramDevirt/import-indir.ll covers the update of
  FunctionSummaryTraits, and test both {def, decl}
- Assembler/thinlto-summary tests both {def, decl}
- The rest of test cases are backfill'ed with def and doesn't test decl.